### PR TITLE
codeintel: Paginate configuration policies

### PR DESCRIPTION
**Pull request by efritz**
This PR adds pagination to configuration policies. In the first draft we just returned an unlimited set of data from the database. This can have pretty adverse effects for large number of policies and I wanted to get this in before branch cut in case customers really try to auto-index _everything_ at once. This should make the UI and worker stable against some kinds of large loads.

Screen Shot 2021-11-18 at 2 56 25 PM
Screen Shot 2021-11-18 at 2 56 34 PM
Screen Shot 2021-11-18 at 2 57 05 PM
Screen Shot 2021-11-18 at 2 56 54 PM

**Backend changes**:

- Rewrote `GetConfigurationPolicies` to take limit, offset, and search conditions.
- Update upload expirer to use new function. Instead of combining all global policies and repository-defined policies, we ask each repository for all policies that apply to them and iterate those.
- Update index scheduler in the same way.
- Update the GraphQL layer to return a policy connection resolver instead of a list to enable pagination.

**Frontend changes**:

- Break apart tabs in configuration policy page so that we have one repository configuration view and one site-admin configuration view
- Use filtered connection to display policies (😓)
- Move delete button into configuration policy page so it matches the flow of indexes/upload records

**Remaining work**:

- [x] Fixing navigation back to the list view when deleting a configuration policy
- [x] Fixing navigation back to the list view when updating a configuration policy

### CODE REVIEW

File: enterprise/cmd/worker/internal/codeintel/janitor/upload_expirer_test.go [hunk 1]
```diff
@@ -135,20 +135,14 @@ func testUploadExpirerMockDBStore(now time.Time) *MockDBStore {
...
-	getConfigurationPolicies := func(ctx context.Context, opts dbstore.GetConfigurationPoliciesOptions) (filtered []dbstore.ConfigurationPolicy, _ error) {
+	getConfigurationPolicies := func(ctx context.Context, opts dbstore.GetConfigurationPoliciesOptions) (filtered []dbstore.ConfigurationPolicy, _ int, _ error) {
 		for _, policy := range policies {
-			if opts.RepositoryID == 0 {
-				if policy.RepositoryID != nil {
-					continue
-				}
-			} else if policy.RepositoryID == nil || *policy.RepositoryID != opts.RepositoryID {
-				continue
+			if policy.RepositoryID == nil || *policy.RepositoryID == opts.RepositoryID {
```
Comment (Strum355) [inline enterprise/cmd/worker/internal/codeintel/janitor/upload_expirer_test.go:140]: previously policy.RepositoryID == nil would result in a skip. Is this change intended?

File: enterprise/internal/codeintel/stores/dbstore/configuration_policies.go [hunk 2]
```diff
@@ -203,9 +218,28 @@ SELECT
...
+// makeConfigurationPolicySearchCondition returns a disjunction of LIKE clauses against all searchable
+// columns of an configuration policy.
+func makeConfigurationPolicySearchCondition(term string) *sqlf.Query {
+	searchableColumns := []string{
+		"p.name",
+	}
+
+	var termConds []*sqlf.Query
+	for _, column := range searchableColumns {
+		termConds = append(termConds, sqlf.Sprintf(column+" ILIKE %s", "%"+term+"%"))
```
Comment (Strum355) [inline enterprise/internal/codeintel/stores/dbstore/configuration_policies.go:237]: Does this scale fine?
image

File: enterprise/internal/codeintel/stores/dbstore/configuration_policies.go [hunk 3]
```diff
@@ -126,65 +115,91 @@ type GetConfigurationPoliciesOptions struct {
...
 		conds = append(conds, sqlf.Sprintf("p.retention_enabled"))
 	}
 	if opts.ForIndexing {
 		conds = append(conds, sqlf.Sprintf("p.indexing_enabled"))
 	}
+	if len(conds) == 0 {
+		conds = append(conds, sqlf.Sprintf("TRUE"))
+	}
 
-	authzConds, err := database.AuthzQueryConds(ctx, s.Store.Handle().DB())
```
Comment (Strum355) [inline enterprise/internal/codeintel/stores/dbstore/configuration_policies.go:171]: No more authzquery conds? :gohno:

File: enterprise/cmd/worker/internal/codeintel/janitor/upload_expirer_test.go (hunk 1 above)
Comment (efritz) [reply enterprise/cmd/worker/internal/codeintel/janitor/upload_expirer_test.go:140]: Yes, with pagination we also changed what happens if you don't supply a repository. Previously it would return policies that didn't have a repository identifier set, now it returns policies unfiltered by that field.

File: enterprise/internal/codeintel/stores/dbstore/configuration_policies.go (hunk 2 above)
Comment (efritz) [reply enterprise/internal/codeintel/stores/dbstore/configuration_policies.go:237]: If the patterns don't match the the entire reps table it will do a bitmap index scan which should be fine. This is all currently done in a background routine or in admin only UIs for now so it's okay if it's higher latency for now.

I don't think it should add significant db load as it's not run super frequently.

**Review by Strum355 — APPROVED** [review]

File: enterprise/internal/codeintel/stores/dbstore/configuration_policies.go (hunk 3 above)
Comment (efritz) [reply enterprise/internal/codeintel/stores/dbstore/configuration_policies.go:171]: We only use repo ids to say which policies apply to it, there's no way to get repo information _from_ the policy (except by looking at patterns or ids).

I'm not sure what to say for things like "yeah this repo matching this pattern you can't see, but it also has two other patterns that match other repos". Leaking the patterns doesn't necessarily indicate the existence of a repository.
